### DISCUSSION TITLE
Tatin packages without tatin server

### DISCUSSION THREAD

**Discussion by kdtp** [aplteam/Tatin#142] [Q&A] [answered]
It all started when I want to install cider. Cause the tatin server is blocked by  a firewall, I tried to install it from a package file. I had to build this zip file (from another machine) first. Now installing from this package file, I learned that it (obviously) looks for its dependencies using the blocked tatin server. Next try: I build all dependencies. And list them with Tatin.InstallPackages. Again it tried to access the tatin server, though it is not needed anymore. Next try: I tried to install them one by one. This worked for packages with no dependencies. But failed for other packages. It still looks up the tatin server, though the needed package is already installed.
After trying some nasty tricks with Tatin.Cache, I gave up.

**Comment by kdtp** [comment]
Here is my wish list:
1) Download the package zip from tatin server
2) tatin.client should be aware of already installed packages
3) Install from a folder with tatin packages, containing the package and all missing dependencies

**Comment by aplteam** [comment]
Tatin questions all Registries it knows about for packages. By default it knows only about two, and that's the principal Tatin server and the test server. It does not contact the test server because that one has a priority=0, meaning it is not part of the search path:

```
      ]Tatin.ListRegistries
 URI                      Alias               Port  Priority 
 -----------------------  ------------------  ----  -------- 
 https://tatin.dev/       tatin                  0       100 
 https://test.tatin.dev/  tatin-test             0         0 
```

You can either remove https://tatin.dev (and https://test.tatin.dev for that matter) or set its priority to 0. That will prevent Tatin from trying to contact it.

When you want Tatin to look elsewhere, you must make that "elsewhere" known to the Tatin client. For that you need a local Registry. If you don't want to run a Tatin server this can be a simple folder, although running a server has several advantages.

The document "Tatin User Settings" provides all the information required for doing that.

Soon there will be a `]Tatin.Copy <Rgistry> <folder> [package-list]` available, allowing you to copy 

* all versions of all packages
* the last version of all packages
* particular packages

from a given server to a particular folder.

I struggle to understand your wish list:

1. Tatin downloads ZIP files?!
2. Tatin *is* aware of already installed packages?!
3. It does if the folder contains the right stuff:
 
   Every package must be a ZIP files with names following this pattern: `<group>-<name>-<major.minor.patch>`. 

   These files must be saved in a folder named in the same way: `<group>-<name>-<major.minor.patch>\`

<https://github.com/aplteam/Tatin/tree/main/TestServer/Registry/> is such a folder. Note that the sub-folder contains more than just the ZIP file, but that's optional.

**Reply by kdtp** [reply]
Thanks a lot, I will try that.

1) In Maven Central you can download a bundle jar for most packages. This is helpful, if you cannot access the repository via mvn. So I like to download aplteam-cider-XXXXX.zip from tatin server (the package is hosted there anyway). As far as I understand ]Tatin.Copy is what I am looking for (an option to download all dependencies, would be nice)
2) I am not sure if I am right or not. But it seems that tatin.client is not checking if a package is already installed before contacting tatin.server.

**Reply by aplteam** [reply]
I overlooked the last bit, sorry:

When you specify a precise package [Registry-alias](group+package-name+full-version-number), and the package has been loaded or installed before and is therefore cached, Tatin does NOT contact any Registry.

However, if you do not specify the group name or the version number, or a part of the version number, Tatin MUST contact every Registry defined in the configuration in order to establish the missing piece(s) of information.

For example:

```
]Tatin.LoadPackages [tatin]aplteam-MarkAPL-12.1.2
```

does not require an Internet connection but

```
]Tatin.LoadPackages aplteam-MarkAPL-12.1.2
```

and

```
]Tatin.LoadPackages [tatin]aplteam-MarkAPL-12.1
```

both do: 

* The first one does not specify a Registry 
* The second one does not specify the patch number

... therefore Tatin must contact all Registries defined in the configuration and ask whether they host that package.

Up to and including version 0.104.0 there was a bug: In case there is only one Registry defined, or at least only one with a priority greater than 0, Tatin should not try to access the Internet even if no Registry is specified by the user, because obviously it will be that Registry in any case.

**Comment by aplteam** [comment, marked as answer]
With version 0.112.0 the user command `]TATIN.CopyRegistry` was introduced. That solves the problem.
